### PR TITLE
feat(server): Add OAuth handlers for GitHub, Notion, Google & amend store data structure

### LINKED ISSUE #7822: Create OAuth handlers for Google, GitHub and Notion
OAuth manager for a service has three jobs:

* Provide a login URL, given a set of scopes
* Exchange an auth code for tokens
* Refresh the auth token

We need an initial implementation for the services determined in #7811.

### PR DESCRIPTION
**Pull request by Pwuts**
- Merge first: #7826
- Resolves #7822

### Changes 🏗️

- Add `BaseOAuthHandler` + 3 initial implementations
  - Add `GitHubOAuthHandler`
  - Add `NotionOAuthHandler`
  - Add `GoogleOAuthHandler`
- Amend `OAuth2Credentials` type
  - Add `metadata` attribute
  - Make `access_token_expires_at`, `refresh_token`, `refresh_token_expires_at` optional

Also:
- Sorted the dependencies in `pyproject.toml`

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [x] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [x] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: rnd/autogpt_server/autogpt_server/integrations/google.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        flow = self._setup_oauth_flow(None)
+        flow.redirect_uri = self.redirect_uri
+        flow.fetch_token(code=code)
+
+        google_creds = flow.credentials
+        # Google's OAuth library is poorly typed so we need some of these:
+        assert google_creds.token
+        assert google_creds.refresh_token
+        assert google_creds.expiry
+        assert google_creds.scopes
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/integrations/google.py:44]: I don't like this. If anyone knows a better way, lmk.

Comment (majdyz) [reply rnd/autogpt_server/autogpt_server/integrations/google.py:44]: Add  `if not google_creds: raise` ?

Comment (Pwuts) [reply rnd/autogpt_server/autogpt_server/integrations/google.py:44]: `google_creds` is definitely set. It just has uncertainly typed attributes.

**Review by majdyz — APPROVED** [review]
I see a big chunks of the code between the handler is replicated, is there a way to put the helper or shared code on the parent abstract function and reuse ?

**Comment by Pwuts** [comment]
@majdyz what do you consider duplicated code? Every provider requires a slightly different implementation which is why these separate handlers exist.

**Review by Torantulino — APPROVED** [review]

**Pull request merged**